File: ui/web-app/src/domain/resourceIndexAdapters.ts
import type { ChartFamilyId, ChartPageData, CurrentArtifactsJson, FlightPlan, MapViewJson, MapViewOptionJson, RegionId, ResourceIndexJson, TileStorageKind } from "./types";

type MapView = MapViewJson;
type MapViewOption = MapViewOptionJson;
type ChartPage = ChartPageData;
type SupportedChartFamily = Extract<ChartFamilyId, "sec" | "tac" | "enr-l" | "enr-h">;
type ChartAsset = ChartPage["airports"][number]["charts"][number];
type FolderCategory = ChartAsset["folder_category"];
type ShadedReliefProduct = NonNullable<CurrentArtifactsJson["static_products"]>[number] & { id: string };

const supportedChartFamilies = new Set<SupportedChartFamily>(["sec", "tac", "enr-l", "enr-h"]);

function isSupportedChartFamily(familyId: ChartFamilyId): familyId is SupportedChartFamily {
  return supportedChartFamilies.has(familyId as SupportedChartFamily);
}

function mapLauncherLabel(familyId: string): string {
  switch (familyId) {
    case "sec":
      return "SEC";
    case "tac":
      return "TAC";
    case "enr-l":
      return "IFR L";
    case "enr-h":
      return "IFR H";
    case "shaded-relief":
      return "RELIEF";
    default:
      return familyId.toUpperCase();
  }
}

function tileStorageKindForCollection(_collectionId: string): TileStorageKind {
  return "sectional_package";
}

function tileUrlRoot(packageName: string): string {
  return `/sectional-packages/${packageName}/tiles`;
}

function tilePathTemplateForCollection(collection: ResourceIndexJson["chart_collections"][number]): string {
  return collection.tile_path_template.replace(/^tiles\//, "");
}

function tileSizeForFamily(_resourceIndex: ResourceIndexJson, _familyId: SupportedChartFamily): number {
  return 512;
}

function minZoomForLevels(levels: MapView["levels"]): number {
  const minLevel = Math.min(...levels.map((level) => level.zoom));
  return Math.max(1.5, minLevel - 2.8);
}

function maxZoomForLevels(levels: MapView["levels"]): number {
  const maxLevel = Math.max(...levels.map((level) => level.zoom));
  return maxLevel + 0.8;
}

function familyDisplayName(resourceIndex: ResourceIndexJson, familyId: SupportedChartFamily): string {
  return resourceIndex.families.find((entry) => entry.id === familyId)?.display_name ?? familyId;
}

function regionDisplayName(resourceIndex: ResourceIndexJson, regionId: RegionId): string {
  return resourceIndex.regions.find((entry) => entry.id === regionId)?.display_name ?? regionId.toUpperCase();
}

function deriveMapView(
  resourceIndex: ResourceIndexJson,
  collection: ResourceIndexJson["chart_collections"][number] & { family_id: SupportedChartFamily },
): MapView {
  const levels = collection.levels.map((level) => ({
    zoom: level.zoom,
    x_min: level.x_min,
    x_max: level.x_max,
    y_tms_min: level.y_tms_min,
    y_tms_max: level.y_tms_max,
  }));
  return {
    chart_family: collection.family_id,
    chart_name: `${regionDisplayName(resourceIndex, collection.region_id)} ${familyDisplayName(resourceIndex, collection.family_id)}`,
    chart_index: collection.chart_index,
    tile_root: "tiles",
    tile_url_root: tileUrlRoot(collection.package_id),
    tile_path_template: tilePathTemplateForCollection(collection),
    tile_size: tileSizeForFamily(resourceIndex, collection.family_id),
    min_zoom: minZoomForLevels(levels),
    max_zoom: maxZoomForLevels(levels),
    storage_kind: tileStorageKindForCollection(collection.id),
    package_name: collection.package_id,
    initial_viewport: {
      lat: collection.default_view.lat,
      lon: collection.default_view.lon,
      zoom: collection.default_view.zoom,
    },
    levels,
  };
}

function fullPyramidLevels(minZoom: number, maxZoom: number): MapView["levels"] {
  const levels: MapView["levels"] = [];
  for (let zoom = minZoom; zoom <= maxZoom; zoom += 1) {
    const maxTile = (2 ** zoom) - 1;
    levels.push({ zoom, x_min: 0, x_max: maxTile, y_tms_min: 0, y_tms_max: maxTile });
  }
  return levels;
}

function shadedReliefRegionId(productId: string): RegionId | null {
  const regionId = productId.replace(/^shaded-relief-/, "") as RegionId;
  return ["ne", "nc", "nw", "se", "sc", "sw", "ec", "ak", "pac"].includes(regionId) ? regionId : null;
}

function defaultViewportForRegion(resourceIndex: ResourceIndexJson, regionId: RegionId): MapView["initial_viewport"] {
  const referenceCollection =
    resourceIndex.chart_collections.find((collection) => collection.region_id === regionId && collection.family_id === "sec")
    ?? resourceIndex.chart_collections.find((collection) => collection.region_id === regionId);
  return referenceCollection?.default_view ?? { lat: 39, lon: -98, zoom: 4 };
}

function deriveShadedReliefMapView(resourceIndex: ResourceIndexJson, product: ShadedReliefProduct, regionId: RegionId): MapView {
  const levels = fullPyramidLevels(0, 10);
  return {
    chart_family: "shaded-relief",
    chart_name: `${regionDisplayName(resourceIndex, regionId)} Shaded Relief`,
    chart_index: 0,
    tile_root: "tiles",
    tile_url_root: `/shaded-relief-products/${product.id}/tiles`,
    tile_path_template: "{z}/{x}/{y}.webp",
    tile_size: 512,
    min_zoom: 0,
    max_zoom: 10.8,
    storage_kind: "static_product",
    package_name: product.id,
    initial_viewport: defaultViewportForRegion(resourceIndex, regionId),
    levels,
  };
}

function deriveShadedReliefMapViews(resourceIndex: ResourceIndexJson, currentArtifacts?: CurrentArtifactsJson): MapViewOption[] {
  return (currentArtifacts?.static_products ?? [])
    .flatMap((product): Array<{ product: ShadedReliefProduct; regionId: RegionId }> => {
      if (!product.id?.startsWith("shaded-relief-")) {
        return [];
      }
      const regionId = shadedReliefRegionId(product.id);
      return regionId ? [{ product: product as ShadedReliefProduct, regionId }] : [];
    })
    .map(({ product, regionId }) => ({
      id: product.id,
      label: `${regionDisplayName(resourceIndex, regionId)} Shaded Relief`,
      region_id: regionId,
      map_view: deriveShadedReliefMapView(resourceIndex, product, regionId),
    }));
}

export function deriveMapViews(
  resourceIndex: ResourceIndexJson,
  preferredIds: string[],
  currentArtifacts?: CurrentArtifactsJson,
): MapViewOption[] {
  const collections = resourceIndex.chart_collections.flatMap((collection) =>
    isSupportedChartFamily(collection.family_id) ? [{ ...collection, family_id: collection.family_id }] : [],
  );
  const chartMapViews = collections.map((collection) => ({
    id: collection.id,
    label: `${regionDisplayName(resourceIndex, collection.region_id)} ${familyDisplayName(resourceIndex, collection.family_id)}`,
    region_id: collection.region_id,
    map_view: deriveMapView(resourceIndex, collection),
  }));
  const allMapViews = [...chartMapViews, ...deriveShadedReliefMapViews(resourceIndex, currentArtifacts)];
  const selectedCollections =
    preferredIds.length > 0
      ? preferredIds
          .map((id) => allMapViews.find((entry) => entry.id === id))
          .filter((entry): entry is MapViewOption => entry !== undefined)
      : allMapViews;
  return selectedCollections;
}

function airportIdsFromPlan(plan: FlightPlan): string[] {
  const airportIds = new Set<string>();
  if (plan.departure) airportIds.add(plan.departure);
  if (plan.destination) airportIds.add(plan.destination);
  if (plan.alternate) airportIds.add(plan.alternate);
  for (const component of plan.route_components ?? []) {
    if (component.kind === "waypoint" && "Airport" in component.waypoint) {
      airportIds.add(component.waypoint.Airport);
    } else if (component.kind === "procedure") {
      airportIds.add(component.procedure.airport_id);
    }
  }
  return [...airportIds];
}

function folderCategoryForRecord(
  kind: "plate" | "csup",
  record: ResourceIndexJson["plates"][number] | ResourceIndexJson["csups"][number],
): FolderCategory {
  const documentType = kind === "csup" ? "csup" : record.document_type;
  switch (documentType) {
    case "airport_diagram":
      return "airport-diagram";
    case "takeoff_minimums":
    case "alternate_minimums":
    case "minimums":
      return "takeoff-mins";
    case "departure":
      return "departure";
    case "star":
      return "star";
    case "csup":
      return "csup";
    case "approach":
    case "other":
    default:
      return "approach";
  }
}

function chartAssetForRecord(
  airportId: string,
  kind: "plate" | "csup",
  record: ResourceIndexJson["plates"][number] | ResourceIndexJson["csups"][number],
): ChartAsset {
  return {
    id: `${kind}:${airportId}:${record.asset_path.split("/").pop() ?? record.asset_path}`,
    airport_id: airportId,
    package_id: record.package_id,
    label: record.label,
    kind,
    folder_category: folderCategoryForRecord(kind, record),
    source_asset_path: record.asset_path,
    asset_path: record.asset_path,
    asset_url: `/${record.asset_path}`,
    thumbnail_source_path: record.thumbnail_path ?? null,
    thumbnail_path: record.thumbnail_path ?? null,
    thumbnail_url: record.thumbnail_path ? `/${record.thumbnail_path}` : null,
    georef: "georef" in record ? record.georef ?? null : null,
  };
}

export function deriveChartPage(
  resourceIndex: ResourceIndexJson,
  samplePlan: FlightPlan,
): ChartPage {
  const plateById = new Map(resourceIndex.plates.map((record) => [record.id, record]));
  const csupById = new Map(resourceIndex.csups.map((record) => [record.id, record]));
  const airportResourcesByAirportId = new Map(
    resourceIndex.airport_resources.map((entry) => [entry.airport_id, entry]),
  );
  const orderedAirportIds = airportIdsFromPlan(samplePlan);
  const airports = [...orderedAirportIds]
    .map((airportId) => {
      const airportResources = airportResourcesByAirportId.get(airportId);
      if (!airportResources) {
        return null;
      }
      const plates = airportResources.plate_ids
        .map((id) => plateById.get(id))
        .filter((record): record is ResourceIndexJson["plates"][number] => record !== undefined)
        .map((record) => chartAssetForRecord(airportId, "plate", record));
      const csups = airportResources.csup_ids
        .map((id) => csupById.get(id))
        .filter((record): record is ResourceIndexJson["csups"][number] => record !== undefined)
        .map((record) => chartAssetForRecord(airportId, "csup", record));
      const charts = [...plates, ...csups];
      if (charts.length === 0) {
        return null;
      }
      return {
        id: airportId,
        label: airportId,
        charts,
      };
    })
    .filter((airport): airport is ChartPage["airports"][number] => airport !== null);

  return {
    airports,
  };
}

export function launcherLabelForFamily(familyId: string): string {
  return mapLauncherLabel(familyId);
}

File: ui/android-app/app/src/main/java/net/jonh/aerobag/prototype/domain/ResourceIndexAdapters.kt
package net.jonh.aerobag.prototype.domain

import kotlinx.serialization.Serializable

@Serializable
data class WireResourceIndex(
    val schema_version: Int,
    val cycle: String? = null,
    val generated_at_utc: String,
    val families: List<WireResourceFamily>,
    val regions: List<WireCatalogRegion>,
    val packages: List<WireResourcePackage>,
    val chart_collections: List<WireChartCollection>,
    val airports: List<WireResourceAirport>,
    val airport_resources: List<WireAirportResources>,
    val plates: List<WireResourcePlate>,
    val csups: List<WireResourceCsup>,
)

@Serializable
data class WireResourceFamily(
    val id: String,
    val display_name: String,
    val kind: String,
)

@Serializable
data class WireResourcePackage(
    val id: String,
    val family_id: String,
    val region_id: WireRegionId,
    val size_bytes: Long,
    val checksum_sha256: String,
)

@Serializable
data class WireChartCollection(
    val id: String,
    val family_id: WireChartFamilyId,
    val region_id: WireRegionId,
    val package_id: String,
    val chart_index: Int,
    val tile_path_template: String,
    val levels: List<WireChartCollectionLevel>,
    val coverage_bounds: WireCoverageBounds,
    val default_view: WireDefaultView,
)

@Serializable
data class WireChartCollectionLevel(
    val zoom: Int,
    val x_min: Int,
    val x_max: Int,
    val y_tms_min: Int,
    val y_tms_max: Int,
)

@Serializable
data class WireCoverageBounds(
    val lat_min: Double,
    val lat_max: Double,
    val lon_min: Double,
    val lon_max: Double,
)

@Serializable
data class WireDefaultView(
    val lat: Double,
    val lon: Double,
    val zoom: Double,
)

@Serializable
data class WireResourceAirport(
    val id: String,
    val facility_name: String,
    val lat: Double,
    val lon: Double,
    val airport_type: String,
)

@Serializable
data class WireAirportResources(
    val airport_id: String,
    val plate_ids: List<String> = emptyList(),
    val csup_ids: List<String> = emptyList(),
    val package_ids: List<String> = emptyList(),
)

@Serializable
data class WireResourcePlate(
    val id: String,
    val airport_id: String,
    val region_id: WireRegionId,
    val package_id: String,
    val asset_path: String,
    val thumbnail_path: String? = null,
    val label: String,
    val asset_kind: String,
    val document_type: String,
)

@Serializable
data class WireResourceCsup(
    val id: String,
    val airport_id: String,
    val region_id: WireRegionId,
    val package_id: String,
    val asset_path: String,
    val thumbnail_path: String? = null,
    val label: String,
    val asset_kind: String,
    val document_type: String,
)

private fun WireChartFamilyId.toUiMapFamily() = when (this) {
    WireChartFamilyId.Sec -> MapChartFamily.Sec
    WireChartFamilyId.Tac -> MapChartFamily.Tac
    WireChartFamilyId.EnrL -> MapChartFamily.EnrL
    WireChartFamilyId.EnrH -> MapChartFamily.EnrH
}

private fun WireChartFamilyId.toResourceId() = when (this) {
    WireChartFamilyId.Sec -> "sec"
    WireChartFamilyId.Tac -> "tac"
    WireChartFamilyId.EnrL -> "enr-l"
    WireChartFamilyId.EnrH -> "enr-h"
}

private fun WireRegionId.toCode() = when (this) {
    WireRegionId.Ne -> "ne"
    WireRegionId.Nc -> "nc"
    WireRegionId.Nw -> "nw"
    WireRegionId.Se -> "se"
    WireRegionId.Sc -> "sc"
    WireRegionId.Sw -> "sw"
    WireRegionId.Ec -> "ec"
    WireRegionId.Ak -> "ak"
    WireRegionId.Pac -> "pac"
}

private fun regionDisplayName(regions: List<WireCatalogRegion>, regionId: WireRegionId): String =
    regions.firstOrNull { it.id == regionId }?.display_name ?: regionId.toCode().uppercase()

private fun familyDisplayName(families: List<WireResourceFamily>, familyId: String): String =
    families.firstOrNull { it.id == familyId }?.display_name ?: familyId

private fun minZoomForLevels(levels: List<TileLevelAvailability>): Double =
    (levels.minOfOrNull { it.zoom } ?: 4) - 2.8

private fun maxZoomForLevels(levels: List<TileLevelAvailability>): Double =
    (levels.maxOfOrNull { it.zoom } ?: 10) + 0.8

private fun tileSizeForFamily(): Int = 512

fun deriveWireCatalog(resourceIndex: WireResourceIndex): WireCatalog {
    val cycle = resourceIndex.cycle ?: "unknown"
    val supportedFamilies = setOf("sec", "tac", "enr-l", "enr-h")
    val familyById = resourceIndex.families.associateBy { it.id }
    val packageById = resourceIndex.packages.associateBy { it.id }

    val families = resourceIndex.families.mapNotNull { family ->
        val familyId = when (family.id) {
            "sec" -> WireChartFamilyId.Sec
            "tac" -> WireChartFamilyId.Tac
            "enr-l" -> WireChartFamilyId.EnrL
            "enr-h" -> WireChartFamilyId.EnrH
            else -> null
        } ?: return@mapNotNull null
        val maxZoom = resourceIndex.chart_collections
            .filter { it.family_id == familyId }
            .maxOfOrNull { collection -> collection.levels.maxOfOrNull { level -> level.zoom } ?: 0 }
        WireCatalogFamily(
            id = familyId,
            display_name = family.display_name,
            kind = family.kind,
            max_zoom = maxZoom,
            tile_size = if (family.kind == "tiled_raster") tileSizeForFamily() else null,
        )
    }

    val packages = resourceIndex.packages.mapNotNull { pkg ->
        val familyId = when (pkg.family_id) {
            "sec" -> WireChartFamilyId.Sec
            "tac" -> WireChartFamilyId.Tac
            "enr-l" -> WireChartFamilyId.EnrL
            "enr-h" -> WireChartFamilyId.EnrH
            else -> null
        } ?: return@mapNotNull null
        WireCatalogPackage(
            id = WirePackageId(
                region = pkg.region_id,
                family = familyId,
                cycle = cycle,
            ),
            package_name = pkg.id,
            family_id = familyId,
            region_id = pkg.region_id,
            cycle = cycle,
            artifact_kind = "zip",
            relative_url = pkg.id,
            manifest_name = pkg.id,
            size_bytes = pkg.size_bytes,
            checksum_sha256 = pkg.checksum_sha256,
        )
    }

    val charts = resourceIndex.chart_collections.filter { it.family_id.toResourceId() in supportedFamilies }.map { collection ->
        val familyDisplay = familyById[collection.family_id.toResourceId()]?.display_name ?: collection.family_id.toResourceId()
        val regionDisplay = regionDisplayName(resourceIndex.regions, collection.region_id)
        WireChartRecord(
            id = WireChartId(
                family = collection.family_id,
                name = collection.id,
                cycle = cycle,
            ),
            family_id = collection.family_id,
            name = collection.id,
            display_name = "$regionDisplay $familyDisplay",
            cycle = cycle,
            region_ids = listOf(collection.region_id),
            max_zoom = collection.levels.maxOfOrNull { it.zoom } ?: 0,
            tile_path_template = collection.tile_path_template,
        )
    }

    val plates = resourceIndex.plates.mapNotNull { plate ->
        val packageRecord = packageById[plate.package_id] ?: return@mapNotNull null
        WirePlateRecord(
            id = WirePlateId(
                airport_id = plate.airport_id,
                procedure_code = plate.id,
                page = 1,
                cycle = cycle,
            ),
            airport_id = plate.airport_id,
            region_id = plate.region_id,
            cycle = cycle,
            procedure_code = plate.id,
            display_name = plate.label,
            kind = plate.asset_kind,
            georeferenced = true,
            page_count = 1,
            asset_base_path = "${packageRecord.id}/${plate.asset_path.removeSuffix(".png")}",
        )
    }

    val supplements = resourceIndex.csups.mapNotNull { csup ->
        val packageRecord = packageById[csup.package_id] ?: return@mapNotNull null
        WireSupplementRecord(
            airport_id = csup.airport_id,
            region_id = csup.region_id,
            cycle = cycle,
            page_count = 1,
            asset_base_path = "${packageRecord.id}/${csup.asset_path.removeSuffix(".png")}",
        )
    }

    return WireCatalog(
        schema_version = resourceIndex.schema_version,
        cycle = cycle,
        catalog_revision = resourceIndex.generated_at_utc,
        families = families,
        regions = resourceIndex.regions,
        packages = packages,
        charts = charts,
        plates = plates,
        supplements = supplements,
    )
}

fun deriveMapViews(
    resourceIndex: WireResourceIndex,
    preferredIds: List<String>,
): List<MapViewOption> {
    val supported = resourceIndex.chart_collections.filter {
        it.family_id == WireChartFamilyId.Sec ||
            it.family_id == WireChartFamilyId.Tac ||
            it.family_id == WireChartFamilyId.EnrL ||
            it.family_id == WireChartFamilyId.EnrH
    }
    val selected = if (preferredIds.isNotEmpty()) {
        preferredIds.mapNotNull { id -> supported.firstOrNull { it.id == id } }
    } else {
        supported
    }
    return selected.map { collection ->
        val levels = collection.levels.map { level ->
            TileLevelAvailability(
                zoom = level.zoom,
                xMin = level.x_min,
                xMax = level.x_max,
                yTmsMin = level.y_tms_min,
                yTmsMax = level.y_tms_max,
            )
        }
        MapViewOption(
            id = collection.id,
            label = "${regionDisplayName(resourceIndex.regions, collection.region_id)} ${familyDisplayName(resourceIndex.families, collection.family_id.toResourceId())}",
            regionId = collection.region_id.toCode(),
            mapView = MapView(
                chartFamily = collection.family_id.toUiMapFamily(),
                chartName = "${regionDisplayName(resourceIndex.regions, collection.region_id)} ${familyDisplayName(resourceIndex.families, collection.family_id.toResourceId())}",
                chartIndex = collection.chart_index,
                tileRoot = "tiles",
                tileUrlRoot = "/sectional-packages/${collection.package_id}/tiles",
                tileSize = tileSizeForFamily(),
                minZoom = minZoomForLevels(levels),
                maxZoom = maxZoomForLevels(levels),
                storageKind = TileStorageKind.SectionalPackage,
                packageName = collection.package_id,
                initialViewport = MapViewportSeed(
                    lat = collection.default_view.lat,
                    lon = collection.default_view.lon,
                    zoom = collection.default_view.zoom,
                ),
                levels = levels,
            ),
        )
    }
}

private fun airportIdsFromPlan(plan: FlightPlan): List<String> {
    val result = linkedSetOf<String>()
    plan.departure?.let(result::add)
    plan.destination?.let(result::add)
    plan.alternate?.let(result::add)
    plan.routeComponents.forEach { component ->
        when (component) {
            is RouteComponent.Waypoint -> airportCode(component.waypoint)?.let(result::add)
            is RouteComponent.Procedure -> result.add(component.procedure.airportId)
            is RouteComponent.Airway -> {}
        }
    }
    return result.toList()
}

private fun airportCode(ref: NavRef): String? = when (ref) {
    is NavRef.Airport -> ref.code
    else -> null
}

private fun folderCategory(documentType: String): String = when (documentType) {
    "airport_diagram" -> "airport-diagram"
    "takeoff_minimums", "alternate_minimums", "minimums" -> "takeoff-mins"
    "departure" -> "departure"
    "star" -> "star"
    "csup" -> "csup"
    "approach", "other" -> "approach"
    else -> "approach"
}

private fun folderCategoryRank(category: String): Int = when (category) {
    "airport-diagram" -> 0
    "csup" -> 1
    "takeoff-mins" -> 2
    "approach" -> 3
    "departure" -> 4
    "star" -> 5
    else -> 6
}

private fun chartAsset(
    airportId: String,
    packageId: String,
    kind: String,
    label: String,
    documentType: String,
    assetPath: String,
    thumbnailPath: String?,
): ChartAsset {
    return ChartAsset(
        id = "$kind:$airportId:${assetPath.substringAfterLast('/')}",
        airportId = airportId,
        packageId = packageId,
        label = if (kind == "csup") "CSup" else label,
        kind = kind,
        folderCategory = folderCategory(if (kind == "csup") "csup" else documentType),
        sourceAssetPath = assetPath,
        assetPath = assetPath,
        assetUrl = "/$assetPath",
        thumbnailSourceAssetPath = thumbnailPath,
        thumbnailAssetPath = thumbnailPath,
        thumbnailUrl = thumbnailPath?.let { "/$it" },
    )
}

fun deriveChartPage(
    resourceIndex: WireResourceIndex,
    samplePlan: FlightPlan,
    allowedPackageIds: Set<String>? = null,
): ChartPageFixture {
    val plateById = resourceIndex.plates.associateBy { it.id }
    val csupById = resourceIndex.csups.associateBy { it.id }
    val airportResourcesByAirportId = resourceIndex.airport_resources.associateBy { it.airport_id }
    val airportIds = linkedSetOf<String>()
    airportIdsFromPlan(samplePlan).forEach(airportIds::add)
    val airports = airportIds.mapNotNull { airportId ->
        val airportResources = airportResourcesByAirportId[airportId] ?: return@mapNotNull null
        val charts = buildList {
            airportResources.plate_ids.mapNotNull(plateById::get).filter { record ->
                allowedPackageIds == null || allowedPackageIds.contains(record.package_id)
            }.forEach { record ->
                add(chartAsset(airportId, record.package_id, "plate", record.label, record.document_type, record.asset_path, record.thumbnail_path))
            }
            airportResources.csup_ids.mapNotNull(csupById::get).filter { record ->
                allowedPackageIds == null || allowedPackageIds.contains(record.package_id)
            }.forEach { record ->
                add(chartAsset(airportId, record.package_id, "csup", record.label, record.document_type, record.asset_path, record.thumbnail_path))
            }
        }.sortedWith(compareBy<ChartAsset>({ folderCategoryRank(it.folderCategory) }, { it.label }))
        if (charts.isEmpty()) {
            null
        } else {
            ChartAirport(
                id = airportId,
                label = airportId,
                charts = charts,
            )
        }
    }
    return ChartPageFixture(
        airports = airports,
    )
}
